File: packages/gh-tasks/src/commands/standup.ts
import { resolveLocale, t } from '../i18n/index.ts';
import type { AppConfig } from '../lib/config.ts';
import { createClient, type GraphQLClient, resolveToken } from '../lib/github.ts';
import { ProjectError, type ProjectRef, resolveProjectRef } from '../lib/project.ts';
import {
  GET_ORG_PROJECT_V2,
  GET_USER_PROJECT_V2,
  GET_VIEWER_LOGIN,
  type GetOrgProjectV2Response,
  type GetUserProjectV2Response,
  type GetViewerLoginResponse,
  LIST_CLOSED_ISSUES,
  LIST_MERGED_PRS,
  LIST_PROJECT_V2_ITEMS,
  LIST_REPO_ISSUES,
  type ListClosedIssuesResponse,
  type ListMergedPRsResponse,
  type ListProjectV2ItemsResponse,
  type ListRepoIssuesResponse,
  type ProjectV2FieldValue,
  type ProjectV2ItemNode,
} from '../lib/queries/index.ts';
import { resolveRepo } from '../lib/repo.ts';
import { detectScope, type Scope } from '../lib/scope.ts';

export interface StandupCommandDeps {
  client?: GraphQLClient;
  hasGitRemote?: () => boolean;
  getRemoteUrl?: () => string | null;
  stdout?: NodeJS.WritableStream;
  stderr?: NodeJS.WritableStream;
  now?: () => Date;
  config?: AppConfig;
}

const FETCH_LIMIT = 100;

/**
 * Standup activity summary.
 *
 * - `--scope repo` (default in a Git repo): Issues / PRs closed-or-merged
 *   since `--since` (default: 24h ago) under "Yesterday", recently-updated
 *   open Issues under "Today", and a placeholder hint under "Blockers".
 * - `--scope org|user`: Projects v2 items whose `updatedAt` is at-or-after
 *   `--since`. Items are split by Status: "Done" → Yesterday, anything
 *   else (incl. unset Status) → Today. Blockers reuses the placeholder
 *   from repo scope — Projects v2 has no "blocked" semantics we can
 *   detect generically.
 *
 * `--mine` filters to items where the viewer is either the content's
 * `author` or one of its `assignees`. DraftIssues have no author/assignee
 * fields, so under `--mine` they are excluded entirely.
 */
export async function standup(
  argv: readonly string[],
  deps: StandupCommandDeps = {}
): Promise<number> {
  const stdout = deps.stdout ?? process.stdout;
  const stderr = deps.stderr ?? process.stderr;
  const locale = resolveLocale(argv, process.env, deps.config);

  const scope = detectScope({ argv, hasGitRemote: deps.hasGitRemote, config: deps.config });
  const since = parseSince(argv, deps.now ? deps.now() : new Date());
  const mine = argv.includes('--mine');

  if (scope === 'repo') {
    return await standupRepo({ argv, deps, locale, stdout, since, mine });
  }

  return await standupProject({ argv, deps, locale, scope, stdout, stderr, since, mine });
}

interface StandupRepoContext {
  argv: readonly string[];
  deps: StandupCommandDeps;
  locale: 'ja' | 'en';
  stdout: NodeJS.WritableStream;
  since: number;
  mine: boolean;
}

async function standupRepo(ctx: StandupRepoContext): Promise<number> {
  const { argv, deps, locale, stdout, since, mine } = ctx;
  const repo = resolveRepo({ argv, getRemoteUrl: deps.getRemoteUrl });
  const client = deps.client ?? createClient(resolveToken());

  const viewerLogin = mine
    ? (await client.request<GetViewerLoginResponse>(GET_VIEWER_LOGIN, {})).viewer.login
    : null;

  const [issuesData, prsData, openIssuesData] = await Promise.all([
    client.request<ListClosedIssuesResponse>(LIST_CLOSED_ISSUES, {
      owner: repo.owner,
      name: repo.name,
      first: FETCH_LIMIT,
    }),
    client.request<ListMergedPRsResponse>(LIST_MERGED_PRS, {
      owner: repo.owner,
      name: repo.name,
      first: FETCH_LIMIT,
    }),
    client.request<ListRepoIssuesResponse>(LIST_REPO_ISSUES, {
      owner: repo.owner,
      name: repo.name,
      first: FETCH_LIMIT,
    }),
  ]);

  const matchesViewer = (n: {
    author?: { login: string } | null;
    assignees?: { nodes: Array<{ login: string }> };
  }): boolean => {
    if (!viewerLogin) return true;
    if (n.author?.login === viewerLogin) return true;
    return n.assignees?.nodes.some((a) => a.login === viewerLogin) ?? false;
  };

  const closedIssues =
    issuesData.repository?.issues.nodes
      .filter((n) => new Date(n.closedAt).getTime() >= since)
      .filter(matchesViewer) ?? [];
  const mergedPRs =
    prsData.repository?.pullRequests.nodes
      .filter((n) => new Date(n.mergedAt).getTime() >= since)
      .filter(matchesViewer) ?? [];
  const openIssues =
    openIssuesData.repository?.issues.nodes
      .filter((n) => new Date(n.updatedAt).getTime() >= since)
      .filter(matchesViewer) ?? [];

  const lines: string[] = [];
  lines.push(`# ${t(locale, 'standup.heading')}${mine && viewerLogin ? ` (@${viewerLogin})` : ''}`);
  lines.push(`since ${new Date(since).toISOString()}`);
  lines.push('');
  lines.push(`## ${t(locale, 'standup.yesterday')}`);
  if (closedIssues.length === 0 && mergedPRs.length === 0) {
    lines.push(`- ${t(locale, 'standup.none')}`);
  } else {
    for (const i of closedIssues) lines.push(`- closed: #${i.number} ${i.title} (${i.url})`);
    for (const p of mergedPRs) lines.push(`- merged: #${p.number} ${p.title} (${p.url})`);
  }
  lines.push('');
  lines.push(`## ${t(locale, 'standup.today')}`);
  if (openIssues.length === 0) {
    lines.push(`- ${t(locale, 'standup.none')}`);
  } else {
    for (const i of openIssues) lines.push(`- in-progress: #${i.number} ${i.title} (${i.url})`);
  }
  lines.push('');
  lines.push(`## ${t(locale, 'standup.blockers')}`);
  lines.push(`- ${t(locale, 'standup.blockersHint')}`);
  lines.push('');
  stdout.write(lines.join('\n'));
  return 0;
}

interface StandupProjectContext {
  argv: readonly string[];
  deps: StandupCommandDeps;
  locale: 'ja' | 'en';
  scope: Exclude<Scope, 'repo'>;
  stdout: NodeJS.WritableStream;
  stderr: NodeJS.WritableStream;
  since: number;
  mine: boolean;
}

async function standupProject(ctx: StandupProjectContext): Promise<number> {
  const { argv, deps, locale, scope, stdout, stderr, since, mine } = ctx;

  let projectRef: ProjectRef;
  try {
    projectRef = resolveProjectRef({ scope, argv, config: deps.config });
  } catch (err) {
    if (err instanceof ProjectError) {
      stderr.write(`${err.message}\n`);
      return 2;
    }
    throw err;
  }

  const client = deps.client ?? createClient(resolveToken());

  const viewerLogin = mine
    ? (await client.request<GetViewerLoginResponse>(GET_VIEWER_LOGIN, {})).viewer.login
    : null;

  const projectId = await resolveProjectNodeId({ client, scope, projectRef });
  if (!projectId) {
    stderr.write(
      `project not found: ${projectRef.owner}/${projectRef.number} (--scope ${scope})\n`
    );
    return 1;
  }

  const data = await client.request<ListProjectV2ItemsResponse>(LIST_PROJECT_V2_ITEMS, {
    projectId,
    first: FETCH_LIMIT,
  });
  if (!data.node) {
    stderr.write(
      `project not found: ${projectRef.owner}/${projectRef.number} (--scope ${scope})\n`
    );
    return 1;
  }

  const inRange = data.node.items.nodes.filter((item) => {
    const ms = new Date(item.updatedAt).getTime();
    return Number.isFinite(ms) && ms >= since;
  });
  const visible = viewerLogin
    ? inRange.filter((item) => matchesViewerOnItem(item, viewerLogin))
    : inRange;

  const yesterday: ProjectV2ItemNode[] = [];
  const today: ProjectV2ItemNode[] = [];
  for (const item of visible) {
    if (isDone(item)) yesterday.push(item);
    else today.push(item);
  }

  const lines: string[] = [];
  lines.push(`# ${t(locale, 'standup.heading')}${mine && viewerLogin ? ` (@${viewerLogin})` : ''}`);
  lines.push(`since ${new Date(since).toISOString()}`);
  lines.push('');
  lines.push(`## ${t(locale, 'standup.yesterday')}`);
  if (yesterday.length === 0) {
    lines.push(`- ${t(locale, 'standup.empty.project')}`);
  } else {
    for (const item of yesterday) lines.push(`- done: ${formatItemLine(item)}`);
  }
  lines.push('');
  lines.push(`## ${t(locale, 'standup.today')}`);
  if (today.length === 0) {
    lines.push(`- ${t(locale, 'standup.empty.project')}`);
  } else {
    for (const item of today) lines.push(`- in-progress: ${formatItemLine(item)}`);
  }
  lines.push('');
  lines.push(`## ${t(locale, 'standup.blockers')}`);
  lines.push(`- ${t(locale, 'standup.blockersHint')}`);
  lines.push('');
  stdout.write(lines.join('\n'));
  return 0;
}

/**
 * Match an item against the viewer login. DraftIssues have no author /
 * assignee fields, so under `--mine` they cannot match — this returns
 * false for them by construction.
 */
function matchesViewerOnItem(item: ProjectV2ItemNode, viewerLogin: string): boolean {
  const c = item.content;
  if (!c) return false;
  if (c.__typename === 'DraftIssue') return false;
  if (c.author?.login === viewerLogin) return true;
  return c.assignees?.nodes.some((a) => a.login === viewerLogin) ?? false;
}

/** Status is "Done" (case-insensitive). Status field is conventionally named "Status". */
function isDone(item: ProjectV2ItemNode): boolean {
  const status = findStatus(item.fieldValues.nodes);
  if (status === null) return false;
  return status.toLowerCase() === 'done';
}

function findStatus(values: readonly ProjectV2FieldValue[]): string | null {
  for (const v of values) {
    if (
      v.__typename === 'ProjectV2ItemFieldSingleSelectValue' &&
      v.field.name.toLowerCase() === 'status'
    ) {
      return v.name;
    }
  }
  return null;
}

function formatItemLine(item: ProjectV2ItemNode): string {
  const c = item.content;
  if (!c) return '(no content)';
  if (c.__typename === 'Issue' || c.__typename === 'PullRequest') {
    const prefix = c.__typename === 'PullRequest' ? 'PR' : '';
    return `${prefix}#${c.number} ${c.title} (${c.url})`;
  }
  return `(draft) ${c.title}`;
}

interface ResolveProjectNodeIdOptions {
  client: GraphQLClient;
  scope: Exclude<Scope, 'repo'>;
  projectRef: ProjectRef;
}

async function resolveProjectNodeId(opts: ResolveProjectNodeIdOptions): Promise<string | null> {
  const { client, scope, projectRef } = opts;
  if (scope === 'org') {
    const data = await client.request<GetOrgProjectV2Response>(GET_ORG_PROJECT_V2, {
      login: projectRef.owner,
      number: projectRef.number,
    });
    return data.organization?.projectV2?.id ?? null;
  }
  const data = await client.request<GetUserProjectV2Response>(GET_USER_PROJECT_V2, {
    login: projectRef.owner,
    number: projectRef.number,
  });
  return data.user?.projectV2?.id ?? null;
}

function parseSince(argv: readonly string[], now: Date): number {
  for (let i = 0; i < argv.length; i++) {
    const arg = argv[i];
    if (arg === undefined) continue;
    if (arg.startsWith('--since=')) {
      const ms = Date.parse(arg.slice('--since='.length));
      if (Number.isFinite(ms)) return ms;
    }
    if (arg === '--since') {
      const next = argv[i + 1];
      if (next !== undefined) {
        const ms = Date.parse(next);
        if (Number.isFinite(ms)) return ms;
      }
    }
  }
  return now.getTime() - 24 * 60 * 60 * 1000;
}

File: cmd/standup.go
package cmd

import (
	"context"
	"errors"
	"fmt"
	"strings"
	"time"

	"github.com/spf13/cobra"

	"github.com/ozzy-labs/gh-tasks/internal/github/queries"
	"github.com/ozzy-labs/gh-tasks/internal/project"
	"github.com/ozzy-labs/gh-tasks/internal/projectitem"
	"github.com/ozzy-labs/gh-tasks/internal/repo"
	"github.com/ozzy-labs/gh-tasks/internal/scope"
)

const standupFetchLimit = 100

func newStandupCmd(deps Deps) *cobra.Command {
	c := &cobra.Command{
		Use:   "standup",
		Short: "Generate standup summary of recent activity",
		RunE: func(c *cobra.Command, _ []string) error {
			return runStandup(c.Context(), c, deps)
		},
	}
	c.Flags().String("since", "", "ISO-8601 timestamp to anchor the window (default: 24h ago)")
	c.Flags().Bool("mine", false, "filter to items where the viewer is author or assignee")
	return c
}

func runStandup(ctx context.Context, c *cobra.Command, deps Deps) error {
	r, err := deps.Resolve(c)
	if err != nil {
		return localizedError(c, r, err)
	}
	sc, err := scope.Detect(scope.DetectOptions{
		Flag:         flagString(c, "scope"),
		HasGitRemote: deps.HasGitRemote,
		DefaultScope: r.Config.DefaultScope,
	})
	if err != nil {
		return localizedError(c, r, err)
	}
	since, err := standupSince(c, deps.Now())
	if err != nil {
		return localizedError(c, r, err)
	}
	mine, _ := c.Flags().GetBool("mine")
	if sc == scope.Repo {
		return runStandupRepo(ctx, c, deps, r, since, mine)
	}
	return runStandupProject(ctx, c, deps, r, sc, since, mine)
}

func runStandupRepo(ctx context.Context, c *cobra.Command, deps Deps, r Resolved, since time.Time, mine bool) error {
	id, err := repo.Resolve(repo.ResolveOptions{Flag: flagString(c, "repo"), GetRemoteURL: deps.GetRemoteURL})
	if err != nil {
		return localizedError(c, r, err)
	}
	clients, err := deps.NewClients()
	if err != nil {
		return localizedError(c, r, err)
	}
	viewerLogin := ""
	if mine {
		v, err := queries.GetViewerLogin(ctx, clients.AsGenqlientClient())
		if err != nil {
			return wrapTransport(c.ErrOrStderr(), r.Locale, "get viewer login", err)
		}
		if v == nil || v.Viewer == nil || v.Viewer.Login == "" {
			return localizedError(c, r, newRuntimeError("error.standup.viewerLoginUnresolved"))
		}
		viewerLogin = v.Viewer.Login
	}
	gqlClient := clients.AsGenqlientClient()
	closedIssues, err := queries.PaginateClosedIssues(ctx, gqlClient, id.Owner, id.Name, standupFetchLimit)
	if errors.Is(err, queries.ErrRepoNotFound) {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.repo.notFound", "owner", id.Owner, "name", id.Name))
		return ErrSilentRuntime
	}
	if err != nil {
		return wrapTransport(c.ErrOrStderr(), r.Locale, "list closed issues", err)
	}
	mergedPRs, err := queries.PaginateMergedPRs(ctx, gqlClient, id.Owner, id.Name, standupFetchLimit)
	if errors.Is(err, queries.ErrRepoNotFound) {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.repo.notFound", "owner", id.Owner, "name", id.Name))
		return ErrSilentRuntime
	}
	if err != nil {
		return wrapTransport(c.ErrOrStderr(), r.Locale, "list merged PRs", err)
	}
	openIssues, err := queries.PaginateRepoIssues(ctx, gqlClient, id.Owner, id.Name, standupFetchLimit)
	if errors.Is(err, queries.ErrRepoNotFound) {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.repo.notFound", "owner", id.Owner, "name", id.Name))
		return ErrSilentRuntime
	}
	if err != nil {
		return wrapTransport(c.ErrOrStderr(), r.Locale, "list repo issues", err)
	}

	type closedItem struct {
		Number int
		Title  string
		URL    string
	}
	closed := []closedItem{}
	for _, n := range closedIssues {
		if n == nil {
			continue
		}
		closedAt := ""
		if n.ClosedAt != nil {
			closedAt = *n.ClosedAt
		}
		if !timeAtOrAfter(closedAt, since) {
			continue
		}
		authorLogin := ""
		if n.Author != nil && *n.Author != nil {
			authorLogin = (*n.Author).GetLogin()
		}
		var assignees []string
		if n.Assignees != nil {
			assignees = make([]string, 0, len(n.Assignees.Nodes))
			for _, a := range n.Assignees.Nodes {
				if a != nil {
					assignees = append(assignees, a.Login)
				}
			}
		}
		if !matchesViewer(authorLogin, assignees, viewerLogin) {
			continue
		}
		closed = append(closed, closedItem{Number: n.Number, Title: n.Title, URL: n.Url})
	}
	type mergedItem struct {
		Number int
		Title  string
		URL    string
	}
	merged := []mergedItem{}
	for _, n := range mergedPRs {
		if n == nil {
			continue
		}
		mergedAt := ""
		if n.MergedAt != nil {
			mergedAt = *n.MergedAt
		}
		if !timeAtOrAfter(mergedAt, since) {
			continue
		}
		authorLogin := ""
		if n.Author != nil && *n.Author != nil {
			authorLogin = (*n.Author).GetLogin()
		}
		var assignees []string
		if n.Assignees != nil {
			assignees = make([]string, 0, len(n.Assignees.Nodes))
			for _, a := range n.Assignees.Nodes {
				if a != nil {
					assignees = append(assignees, a.Login)
				}
			}
		}
		if !matchesViewer(authorLogin, assignees, viewerLogin) {
			continue
		}
		merged = append(merged, mergedItem{Number: n.Number, Title: n.Title, URL: n.Url})
	}
	type openItem struct {
		Number int
		Title  string
		URL    string
	}
	open := []openItem{}
	for _, n := range openIssues {
		if n == nil {
			continue
		}
		if !timeAtOrAfter(n.UpdatedAt, since) {
			continue
		}
		authorLogin := ""
		if n.Author != nil && *n.Author != nil {
			authorLogin = (*n.Author).GetLogin()
		}
		var assignees []string
		if n.Assignees != nil {
			assignees = make([]string, 0, len(n.Assignees.Nodes))
			for _, a := range n.Assignees.Nodes {
				if a != nil {
					assignees = append(assignees, a.Login)
				}
			}
		}
		if !matchesViewer(authorLogin, assignees, viewerLogin) {
			continue
		}
		open = append(open, openItem{Number: n.Number, Title: n.Title, URL: n.Url})
	}

	out := c.OutOrStdout()
	header := r.T("standup.heading")
	if mine && viewerLogin != "" {
		header = fmt.Sprintf("%s (@%s)", header, viewerLogin)
	}
	fmt.Fprintf(out, "# %s\n", header)
	fmt.Fprintf(out, "since %s\n\n", since.UTC().Format(time.RFC3339))
	fmt.Fprintf(out, "## %s\n", r.T("standup.yesterday"))
	if len(closed) == 0 && len(merged) == 0 {
		fmt.Fprintf(out, "- %s\n", r.T("standup.none"))
	} else {
		for _, i := range closed {
			fmt.Fprintf(out, "- closed: #%d %s (%s)\n", i.Number, i.Title, i.URL)
		}
		for _, p := range merged {
			fmt.Fprintf(out, "- merged: #%d %s (%s)\n", p.Number, p.Title, p.URL)
		}
	}
	fmt.Fprintf(out, "\n## %s\n", r.T("standup.today"))
	if len(open) == 0 {
		fmt.Fprintf(out, "- %s\n", r.T("standup.none"))
	} else {
		for _, i := range open {
			fmt.Fprintf(out, "- in-progress: #%d %s (%s)\n", i.Number, i.Title, i.URL)
		}
	}
	fmt.Fprintf(out, "\n## %s\n- %s\n", r.T("standup.blockers"), r.T("standup.blockersHint"))
	return nil
}

func runStandupProject(ctx context.Context, c *cobra.Command, deps Deps, r Resolved, sc scope.Scope, since time.Time, mine bool) error {
	pref, err := project.Resolve(project.ResolveOptions{
		Scope:       sc,
		Flag:        flagString(c, "project"),
		OrgProject:  r.Config.OrgProject,
		UserProject: r.Config.UserProject,
	})
	if err != nil {
		return localizedError(c, r, err)
	}
	clients, err := deps.NewClients()
	if err != nil {
		return localizedError(c, r, err)
	}
	viewerLogin := ""
	if mine {
		v, err := queries.GetViewerLogin(ctx, clients.AsGenqlientClient())
		if err != nil {
			return wrapTransport(c.ErrOrStderr(), r.Locale, "get viewer login", err)
		}
		if v == nil || v.Viewer == nil || v.Viewer.Login == "" {
			return localizedError(c, r, newRuntimeError("error.standup.viewerLoginUnresolved"))
		}
		viewerLogin = v.Viewer.Login
	}
	pid, err := projectitem.ResolveProjectNodeID(ctx, clients.GraphQL, sc, pref)
	if err != nil {
		return localizedError(c, r, err)
	}
	if pid == "" {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.project.notFound", "owner", pref.Owner, "number", pref.Number, "scope", sc))
		return ErrSilentRuntime
	}
	items, err := queries.PaginateProjectV2Items(ctx, clients.AsGenqlientClient(), pid, standupFetchLimit)
	if errors.Is(err, queries.ErrProjectNotFound) {
		fmt.Fprintln(c.ErrOrStderr(), r.T("error.project.notFound", "owner", pref.Owner, "number", pref.Number, "scope", sc))
		return ErrSilentRuntime
	}
	if err != nil {
		return wrapTransport(c.ErrOrStderr(), r.Locale, "list project items", err)
	}
	yesterday := []*queries.ProjectV2ItemNode{}
	today := []*queries.ProjectV2ItemNode{}
	for _, item := range items {
		if item == nil {
			continue
		}
		if !timeAtOrAfter(item.UpdatedAt, since) {
			continue
		}
		if viewerLogin != "" && !matchesViewerOnItem(item, viewerLogin) {
			continue
		}
		if isItemDone(item) {
			yesterday = append(yesterday, item)
		} else {
			today = append(today, item)
		}
	}
	out := c.OutOrStdout()
	header := r.T("standup.heading")
	if mine && viewerLogin != "" {
		header = fmt.Sprintf("%s (@%s)", header, viewerLogin)
	}
	fmt.Fprintf(out, "# %s\n", header)
	fmt.Fprintf(out, "since %s\n\n", since.UTC().Format(time.RFC3339))
	fmt.Fprintf(out, "## %s\n", r.T("standup.yesterday"))
	if len(yesterday) == 0 {
		fmt.Fprintf(out, "- %s\n", r.T("standup.empty.project"))
	} else {
		for _, item := range yesterday {
			fmt.Fprintf(out, "- done: %s\n", projectitem.FormatItemLineCompact(item))
		}
	}
	fmt.Fprintf(out, "\n## %s\n", r.T("standup.today"))
	if len(today) == 0 {
		fmt.Fprintf(out, "- %s\n", r.T("standup.empty.project"))
	} else {
		for _, item := range today {
			fmt.Fprintf(out, "- in-progress: %s\n", projectitem.FormatItemLineCompact(item))
		}
	}
	fmt.Fprintf(out, "\n## %s\n- %s\n", r.T("standup.blockers"), r.T("standup.blockersHint"))
	return nil
}

func standupSince(c *cobra.Command, now time.Time) (time.Time, error) {
	since, _ := c.Flags().GetString("since")
	if since == "" {
		return now.Add(-24 * time.Hour), nil
	}
	t, err := time.Parse(time.RFC3339, since)
	if err != nil {
		return time.Time{}, newArgError("error.standup.invalidSince", "value", since)
	}
	return t, nil
}

func timeAtOrAfter(iso string, threshold time.Time) bool {
	if iso == "" {
		return false
	}
	t, err := time.Parse(time.RFC3339, iso)
	if err != nil {
		return false
	}
	return t.Equal(threshold) || t.After(threshold)
}

func matchesViewer(authorLogin string, assigneeLogins []string, viewer string) bool {
	if viewer == "" {
		return true
	}
	if authorLogin == viewer {
		return true
	}
	for _, login := range assigneeLogins {
		if login == viewer {
			return true
		}
	}
	return false
}

// matchesViewerOnItem reports whether the project item's content carries
// the given viewer as author or assignee. Draft items have no author /
// assignees so they're always excluded under `--mine`.
func matchesViewerOnItem(item *queries.ProjectV2ItemNode, viewer string) bool {
	c := projectitem.ContentOf(item)
	if c.Typename == "" || c.Typename == "DraftIssue" {
		return false
	}
	if c.Author != "" && c.Author == viewer {
		return true
	}
	for _, login := range c.Assignees {
		if login == viewer {
			return true
		}
	}
	return false
}

func isItemDone(item *queries.ProjectV2ItemNode) bool {
	status := projectitem.FindStatus(projectitem.FieldValuesOf(item))
	if status == "" {
		return false
	}
	return strings.EqualFold(status, "done")
}
